drv/transceivers-server/src/net_api.rs

```diff
@@ -0,0 +1,414 @@
+                            .map_err(|_e| Error::ReadFailed)?;
+                    }
+                    PowerMode::High => {
+                        // Power enabled, LpMode disabled
+                        self.transceivers
+                            .clear_lpmode(mask)
+                            .map_err(|_e| Error::ReadFailed)?;
+                        self.transceivers
+                            .set_power_enable(mask)
+                            .map_err(|_e| Error::ReadFailed)?;
+                    }
+                }
+                Ok((HostResponse::Ack, 0))
+            }
+            HostRequest::ManagementInterface(i) => {
+                todo!()
+            }
+        }
+    }
+
+    fn select_page(
+        &mut self,
+        upper_page: UpperPage,
+        modules: ModuleId,
+    ) -> Result<(), Error> {
+        let mask = get_mask(modules)?;
+        match upper_page {
+            UpperPage::Cmis(page) => {
+                self.transceivers
+                    .set_i2c_write_buffer(&[page.page()])
+                    .map_err(|_e| Error::ReadFailed)?;
+                self.transceivers
+                    .setup_i2c_write(0x7F, 1, mask)
+                    .map_err(|_e| Error::ReadFailed)?;
+                sleep_for(5); // TODO: Poll instead
+
+                if let Some(bank) = page.bank() {
+                    self.transceivers
+                        .set_i2c_write_buffer(&[bank])
+                        .map_err(|_e| Error::ReadFailed)?;
+                    self.transceivers
+                        .setup_i2c_write(0x7E, 1, mask)
+                        .map_err(|_e| Error::ReadFailed)?;
+                    sleep_for(5); // TODO: Poll instead
+                }
+            }
+            UpperPage::Sff8636(page) => {
+                self.transceivers
+                    .set_i2c_write_buffer(&[page.page()])
+                    .map_err(|_e| Error::ReadFailed)?;
+                self.transceivers
+                    .setup_i2c_write(0x7F, 1, mask)
+                    .map_err(|_e| Error::ReadFailed)?;
+                sleep_for(5); // TODO: Poll instead
+            }
+        }
+        Ok(())
+    }
+
+    fn read(
+        &mut self,
+        mem: MemoryRegion,
+        modules: ModuleId,
+        out: &mut [u8],
+    ) -> Result<(), Error> {
+        // Reading beyond a 128-byte page boundary will wrap around; return an
+        // error if that's going to happen.
+        let page_end = (mem.offset() as u32 / 128 + 1) * 128;
+        if mem.offset() as u32 + mem.len() as u32 > page_end {
+            // TODO use a more descriptive error name here?
+            return Err(Error::InvalidMemoryAccess {
+                offset: mem.offset(),
+                len: mem.len(),
+            });
+        }
+        // CMIS spec only guarantee an 8-byte read (?!)
+        // (see CMIS 5.0, 5.2.2.1)
+        if matches!(mem.upper_page(), UpperPage::Cmis(..)) && mem.len() > 8 {
```

Comment:
This check should be obviated by the type system once the rework PR I mentioned above is merged (assuming it _is_ merged 😄).